manager/controlapi/ca_rotation.go

```diff
@@ -0,0 +1,216 @@
+package controlapi
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/cloudflare/cfssl/helpers"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/ca"
+	"github.com/docker/swarmkit/log"
+)
+
+var minRootExpiration = helpers.OneYear
+
+// Creates and updates the api.RootRotation object and the cross-signed intermediate.
+// This function assumes that a root rotation is not already in progress, and that the root
+// cert and key to be rotated in or the external CAs in the cluster have already been validated
+func updateRootRotationObject(ctx context.Context, securityConfig *ca.SecurityConfig, cluster *api.Cluster, newRootCA ca.RootCA, version uint64) error {
+	var (
+		rootCert, rootKey, crossSignedCert []byte
+		newRootHasSigner                   bool
+		err                                error
+	)
+
+	rootCert = newRootCA.Certs
+	if s, err := newRootCA.Signer(); err == nil {
+		rootCert, rootKey = s.Cert, s.Key
+		newRootHasSigner = true
+	}
+
+	// we have to sign with the original signer, not whatever is in the SecurityConfig's RootCA (which may have an intermediate signer, if
+	// a root rotation is already in progress)
+	switch {
+	case len(cluster.RootCA.CAKey) > 0:
```

Comment:
Two questions:
- Isn't there a method on `cluster.RootCA` that allows us to know if it can sign or not (that doesn't involve checking if the Key bytes are `>0`), like `.Signer()` or `.CanSign()`?
- Why are we creating a `ca.NewRootCA`? Can't we simply use `cluster.RootCA.CrossSignCACertificate` and check if that returns an `err`?